coffee/rewriter.py

```diff
@@ -253,6 +210,179 @@ def factorize(self, mode='standard'):
+                  A[j][k] += ...f(B[0]*C[i][0] + B[1]*C[i][1] + ...)...
+        """
+        # Get all loop nests, then discard the one enclosing the expression
+        nests = [n for n in visit(self.expr_info.loops_parents[0])['fors']]
+        injectable_nests = [n for n in nests if zip(*n)[0] != self.expr_info.loops]
+
+        # Full unroll any unrollable, injectable loop
+        for nest in injectable_nests:
+            to_unroll = [(l, p) for l, p in nest if l not in self.expr_info.loops]
+            nest_writers = FindInstances(Writer).visit(to_unroll[0][0])
+            for op, stmts in nest_writers.items():
+                if op in [Assign, IMul, IDiv]:
+                    # Unroll is unsafe, skip
+                    continue
+                # At this point, must be /op = Incr/
+                for stmt in stmts:
+                    sym, expr = stmt.children
+                    if stmt in [l.incr for l, p in to_unroll]:
+                        # Ignore useless stmts
+                        continue
+                    for l, p in reversed(to_unroll):
+                        inject_expr = [dcopy(expr) for i in range(l.size)]
+                        # Update rank of symbols
+                        for i, e in enumerate(inject_expr):
+                            e_syms = FindInstances(Symbol).visit(e)[Symbol]
+                            for s in e_syms:
+                                s.rank = tuple([r if r != l.dim else i for r in s.rank])
+                        expr = ast_make_expr(Sum, inject_expr)
+                    # Inject the unrolled operation into the expression
+                    ast_replace(self.stmt, {sym: expr}, copy=True)
+                    # Clean up
+                    p.children.remove(self.decls[sym.symbol])
+                    self.decls.pop(sym.symbol)
+            for l, p in to_unroll:
+                p.children.remove(l)
+
+    def simplify(self):
+        """Simplify an expression by applying transformations which should enhance
+        the effectiveness of later rewriting passes. The transformations applied
+        are: ::
+
+            * ``division replacement``: replace divisions by a constant with
+                multiplications, which gives expansion and factorization more
+                rewriting opportunities
+            * ``loop reduction``: simplify loops along which a reduction is
+                performed by 1) preaccumulating along the dimension of the
+                reduction, 2) removing the reduction loop, 3) precomputing
+                constant expressions
+        """
+        # Aliases
+        stmt, expr_info = self.stmt, self.expr_info
+
+        # Division replacement
+        divisions = FindInstances(Div).visit(stmt.children[1])[Div]
+        to_replace = {}
+        for i in divisions:
+            if isinstance(i.right, Symbol) and isinstance(i.right.symbol, float):
+                to_replace[i] = Prod(i.left, Div(1, i.right.symbol))
```

Comment:
I don't know if double precision arithmetic in python is identical to that of ANSI C compilers?